lib/iris/analysis/_interpolation.py

```diff
@@ -472,6 +471,18 @@ def _interpolated_dtype(self, dtype):
             result = np.result_type(_DEFAULT_DTYPE, dtype)
         return result
 
+    @staticmethod
+    def _product(dimensions):
+        # Calculate cartesian product using numpy instead of itertools.
+        # inspired by http://stackoverflow.com/posts/11146645/revisions
+        ndim = len(dimensions)
+        shape = [len(points) for points in dimensions] + [ndim]
+        product = np.empty(shape, dtype=dimensions[0].dtype)
```

Comment:
I'm not sure we can assume that the dtype is consistent. See https://github.com/SciTools/iris/pull/1717/files#diff-07b1c9cb6b22e14b7f314da7af652aafR554 which may produce different dtypes for each dimension.